pkg/cloud/baremetal/actuators/machine/actuator.go

```diff
@@ -160,48 +160,73 @@ func (a *Actuator) Delete(ctx context.Context, machine *machinev1beta1.Machine)
 	if err != nil {
 		return err
 	}
-	if host != nil && host.Spec.ConsumerRef != nil {
-		// don't remove the ConsumerRef if it references some other machine
-		if !consumerRefMatches(host.Spec.ConsumerRef, machine) {
-			log.Printf("host associated with %v, not machine %v.",
-				host.Spec.ConsumerRef.Name, machine.Name)
-			return nil
-		}
-		if host.Spec.Image != nil || host.Spec.Online || host.Spec.UserData != nil {
-			host.Spec.Image = nil
-			host.Spec.Online = false
-			host.Spec.UserData = nil
-			err = a.client.Update(ctx, host)
-			if err != nil && !errors.IsNotFound(err) {
-				return err
-			}
-			return &machineapierrors.RequeueAfterError{}
+
+	if host == nil {
+		log.Printf("finished deleting machine %v.", machine.Name)
+		return nil
+	}
+
+	log.Printf("deleting machine %v using host %v", machine.Name, host.Name)
+
+	if host.Spec.ConsumerRef == nil {
+		log.Printf("removing host annotation from machine %v", machine.Name)
+		_, err := a.clearAnnotation(ctx, machine)
+		if err != nil {
+			return err
 		}
+		return nil // updating the machine will re-enqueue it without us asking
+	}
 
-		waiting := true
-		switch host.Status.Provisioning.State {
-		case bmh.StateRegistrationError, bmh.StateRegistering,
-			bmh.StateMatchProfile, bmh.StateInspecting,
-			bmh.StateReady, bmh.StateProvisioningError,
-			bmh.StateUnmanaged:
-			// Host is not provisioned
-			waiting = false
-		case bmh.StateExternallyProvisioned:
-			// We have no control over provisioning, so just wait until the
-			// host is powered off
-			waiting = host.Status.PoweredOn
+	// don't remove the ConsumerRef if it references some other
+	// machine, but remove the annotation linking this machine to the
+	// host so the current machine can continue deleting
+	if !consumerRefMatches(host.Spec.ConsumerRef, machine) {
+		log.Printf("host associated with %v, not machine %v.",
+			host.Spec.ConsumerRef.Name, machine.Name)
+		_, err := a.clearAnnotation(ctx, machine)
+		if err != nil {
+			return err
 		}
-		if waiting {
-			return &machineapierrors.RequeueAfterError{RequeueAfter: requeueAfter}
-		} else {
-			host.Spec.ConsumerRef = nil
-			err = a.client.Update(ctx, host)
-			if err != nil && !errors.IsNotFound(err) {
-				return err
-			}
+		return nil
+	}
+
+	if host.Spec.Image != nil || host.Spec.Online || host.Spec.UserData != nil {
+		log.Printf("starting to deprovision host %v", host.Name)
+		host.Spec.Image = nil
+		host.Spec.Online = false
+		host.Spec.UserData = nil
+		err = a.client.Update(ctx, host)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
 		}
+		return &machineapierrors.RequeueAfterError{}
+	}
+
+	waiting := true
+	switch host.Status.Provisioning.State {
+	case bmh.StateRegistrationError, bmh.StateRegistering,
+		bmh.StateMatchProfile, bmh.StateInspecting,
+		bmh.StateReady, bmh.StateProvisioningError,
+		bmh.StateUnmanaged:
```

Comment:
StateDeleting should also theoretically be on this list, since it indicates that deprovisioning is complete.
It will eventually work in that state, once the Host is actually deleted and we can't find it any more.

Maybe we should flip the default here, and only set `waiting` to be true in the Provisioning, Provisioned, and Deprovisioning states.

Reply:
Yeah, good idea. Done.